sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamCalcRel.java

```diff
@@ -386,24 +386,44 @@ public Expression field(BlockBuilder list, int index, Type storageType) {
       }
       Expression field = Expressions.call(expression, getter, Expressions.constant(index));
       if (fromType.getTypeName().isLogicalType()) {
-        field = Expressions.call(field, "getMillis");
+        Expression millisField = Expressions.call(field, "getMillis");
         String logicalId = fromType.getLogicalType().getIdentifier();
         if (logicalId.equals(TimeType.IDENTIFIER)) {
-          field = Expressions.convert_(field, int.class);
+          field =
+              Expressions.condition(
+                  Expressions.equal(field, Expressions.constant(null)),
+                  Expressions.constant(null),
+                  Expressions.call(
+                      Integer.class, "valueOf", Expressions.convert_(millisField, int.class)));
         } else if (logicalId.equals(DateType.IDENTIFIER)) {
           field =
-              Expressions.convert_(
-                  Expressions.modulo(field, Expressions.constant(MILLIS_PER_DAY)), int.class);
+              Expressions.condition(
+                  Expressions.equal(field, Expressions.constant(null)),
+                  Expressions.constant(null),
+                  Expressions.call(
+                      Integer.class,
+                      "valueOf",
+                      Expressions.convert_(
+                          Expressions.modulo(millisField, Expressions.constant(MILLIS_PER_DAY)),
```

Comment:
@akedin @apilloud Is usage modulo correct here ? I was trying to write a UT here and I think we need to use divide operation.
Can you please share your views.

Reply:
R: @akedin R: @apilloud  to share the views on the above comment and also on the 
handling of null.

Reply:
Yes, I think you found a bug here. I would expect this to divide.

Reply:
I have fixed it now, by making it a division operation.